LineNumbers.md

```diff
@@ -0,0 +1,320 @@
+return "", "", false, ""
+```
+
+```go "Namedblock Regex"
+namedBlockRe = regexp.MustCompile("^`{3,}\\s?(\\w*)\\s*\"(.+)\"\\s*([+][=])?$")
+```
+
+```go "Fileblock Regex"
+fileBlockRe = regexp.MustCompile("^`{3,}\\s?(\\w+)\\s+([\\w\\.\\-\\/]+)\\s*([+][=])?$")
+```
+
+The (actual) replacement in Replace also joins strings, we need to append to
+ret instead. The returned codeblock from the map blocks are replaced with the
+new prefix and "expanded" (...) before appended.
+
+```go "Lookup replacement and add to ret"
+bname := BlockName(matches[2])
+if val, ok := blocks[bname]; ok {
+	ret = append(ret, val.Replace(prefix+matches[1])...)
+} else {
+	fmt.Fprintf(os.Stderr, "Warning: Block named %s referenced but not defined.\n", bname)
+	ret = append(ret, v)
+}
+```
+
+Even lines which aren't replaced need to be appended instead of joining lines.
+While we are at it let's cleanup empty lines when passing through (they should
+not have prefix).
+
+```go "Handle replace line"
+matches := replaceRe.FindStringSubmatch(line)
+if matches == nil {
+	if v.text != "\n" {
+		v.text = prefix + v.text
+	}
+	ret = append(ret, v)
+	continue
+}
+<<<Lookup replacement and add to ret>>>
+```
+
+Finally something we can replace with a simpler implementation. Instead of
+reading line by line from c (a CodeBlock, earlier a long string with newlines),
+we're now able to use range. The string `line` is still analyzed, "de-macro-ed"
+and prefixed.
+
+```go "Replace codeblock implementation"
+var line string
+for _, v := range c {
+	line = v.text
+	<<<Handle replace line>>>
+}
+return
+```
+
+We need to use all the information we gathered, lets add a "Finalize"-function
+which handles CodeLines and outputs the final strings with the line-directives
+we been working for.
+
+```go "other functions" +=
+<<<Finalize Declaration>>>
+```
+
+Let us implement a method which acts upon a codeblock and returns a string
+containing the ready to be used textual representation of the code (or other
+files). We check if the filename is the same as the previous line, and if the
+linenumber has been increased by exactly one. If any of those have changed it
+is because the source for this line are not the same as the previous and we
+interject a "line directive". We know of two variants of these go and the
+common from the C family of languages. lastly we save state, so we have
+something to compare to on the next line.
+
+```go "Finalize Declaration"
+
+// Finalize extract the textual lines from CodeBlocks and (if needed) prepend a
+// notice about "unexpected" filename or line changes, which is extracted from
+// the contained CodeLines. The result is a string with newlines ready to be
+// pasted into a file.
+func (c CodeBlock) Finalize() (ret string) {
+	var file File
+	var formatstring string
+	var linenumber int
+	for _, l := range c {
+		if linenumber+1 != l.number || file != l.file {
+			switch l.lang {
+			case "go", "golang":
+				formatstring = "//line %[2]v:%[1]v\n"
+			default:
```

Comment:
```suggestion
			case "C", "c":
```

I don't think it makes sense to include a global default, it's not a safe assumption that if it's not Go it's C. It's better to just not include line number if it's not a known language.